api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ImmunizationTranslatorImpl.java

```diff
@@ -0,0 +1,215 @@
+		
+		// taking the visit's most recent immunization encounter
+		Optional<Encounter> encounter = visit.getEncounters().stream()
+		        .filter(e -> encounterType.equals(e.getEncounterType()))
+		        .max(Comparator.comparing(Encounter::getEncounterDatetime));
+		
+		final Encounter newEncounter = new Encounter();
+		newEncounter.setVisit(visit);
+		newEncounter.setLocation(location);
+		newEncounter.setEncounterType(encounterType);
+		newEncounter.setPatient(patient);
+		newEncounter.setProvider(encounterRole, provider);
+		if (visit != null && visit.getStopDatetime() != null) {
+			newEncounter.setEncounterDatetime(visit.getStopDatetime());
+		} else {
+			newEncounter.setEncounterDatetime(openMrsImmunization.getObsDatetime());
+		}
+		
+		openMrsImmunization.setPerson(patient);
+		openMrsImmunization.setLocation(location);
+		openMrsImmunization.setEncounter(encounter.orElse(newEncounter));
+		openMrsImmunization.getGroupMembers().stream().forEach(obs -> {
+			obs.setPerson(patient);
+			obs.setLocation(location);
+			obs.setEncounter(encounter.orElse(newEncounter));
+		});
+		
+		Map<String, Obs> members = getObsMembersMap(openMrsImmunization);
+		
+		Coding coding = fhirImmunization.getVaccineCode().getCoding().stream()
+		        .filter(code -> StringUtils.isEmpty(code.getSystem())).reduce((code1, code2) -> {
+			        throw new IllegalArgumentException("Multiple system-less coding found for the immunization's vaccine: "
+			                + code1.getCode() + " and " + code2.getCode()
+			                + ". No unique system concept could be identified as the coded answer.");
+		        }).get();
+		members.get(ciel984).setValueCoded(getConceptService().getConceptByUuid(coding.getCode()));
+		
+		members.get(ciel1410).setValueDatetime(fhirImmunization.getOccurrenceDateTimeType().getValue());
+		
+		if (CollectionUtils.size(fhirImmunization.getProtocolApplied()) != 1) {
+			throw new IllegalArgumentException(
+			        "Either no protocol applied was found or multiple protocols applied were found. Only strictly one protocol is currently supported for each immunization.");
+		}
+		ImmunizationProtocolAppliedComponent protocolApplied = fhirImmunization.getProtocolApplied().get(0);
+		members.get(ciel1418).setValueNumeric(protocolApplied.getDoseNumberPositiveIntType().getValue().doubleValue());
+		
+		members.get(ciel1419).setValueText(fhirImmunization.getManufacturer().getDisplay());
+		
+		members.get(ciel1420).setValueText(fhirImmunization.getLotNumber());
+		
+		members.get(ciel165907).setValueDatetime(fhirImmunization.getExpirationDate());
+		
+		return openMrsImmunization;
+	}
+	
+	@Override
+	public Immunization toFhirResource(Obs openMrsImmunization) {
+		validateImmunizationObsGroup(openMrsImmunization);
+		
+		Immunization immunization = new Immunization();
+		immunization.setId(openMrsImmunization.getUuid());
+		immunization.setStatus(ImmunizationStatus.COMPLETED);
+		immunization.setPatient(
```

Comment:
Here, I would use our methods from `BaseReferenceHandlingTranslator` or use the corresponding  `PatientReferenceTranslator` and `PractitionerReferenceTranslator` classes as well as the `VisitReferenceTranslator` class you added. While there's nothing wrong with this, it does help ensure some consistency in what we're populating (e.g., our default implementations of patient references and practitioner references include the `display` field, which in turn helps populate the narrative, etc).

Reply:
Addressed [here](https://github.com/openmrs/openmrs-module-fhir2/pull/264/files#diff-56eb3743524091606fd9093205f0bc07R189-R192).